test/hotspot/jtreg/serviceability/sa/TestJhsdbJstackUpcall.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import jdk.test.lib.JDKToolLauncher;
+import jdk.test.lib.SA.SATestUtils;
+import jdk.test.lib.Utils;
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.process.OutputAnalyzer;
+
+/**
+ * @test
+ * @bug 8339307
+ * @requires vm.hasSA
+ * @library /test/lib
+ * @run driver TestJhsdbJstackUpcall
+ */
+public class TestJhsdbJstackUpcall {
+
+    private static final int MAX_ITERATIONS = 20;
+
+    /*
+     * Test should focus JNI call (caller of upcall) because the frame
+     * prior to the upcall cannot be obtained if some exception happens
+     * in during to process upcall.
+     */
+    private static boolean isJNIFrame(List<String> lines) {
+        return lines.stream()
+                    .anyMatch(s -> s.startsWith(" - LingeredAppWithFFMUpcall.callJNI"));
+    }
+
+    private static void runJstackInLoop(LingeredApp app) throws Exception {
+        for (int i = 0; i < MAX_ITERATIONS; i++) {
```

Comment:
What is the reason for doing 20 iterations. Is it because you are waiting for THREAD_NAME to enter the sleep() call? If so, we've addressed this in the past for the general case of wanting to do a "stable"stack trace by using the LingeredApp's SteadyStateThread. LingeredApp.startApp() will not return until this thread has become stable (blocked). Maybe you can do something similar with THREAD_NAME.

Reply:
TBH this testcase is based on TestJhsdbJstackMixed.java , so I'm not stick this code. I will fix to use SteadyStateThread.

Reply:
Ok. I think TestJhsdbJstackMixed has a loop because it's trying to get a stack trace at a point when the thread is executing at a PC that used result in an SA exception (which is now fixed).

SteadyStateThread might be a bit tricky to integrate into this test since the test has no control over what the thread does. Possibly with a new LIngeredApp subclass you could control SteadyStateThread to call your FFM code.